{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial: KEGG REST API commands in KEGGutils\n",
    "## How to make basic requests to KEGG API using intuitive function calls from within your code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import KEGGutils as kg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "print(kg.keggapi_info.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infos on hsa from KEGG:\n",
      "\n",
      "T01001           Homo sapiens (human) KEGG Genes Database\n",
      "hsa              Release 89.0+/03-20, Mar 19\n",
      "                 Kanehisa Laboratories\n",
      "                 38,683 entries\n",
      "\n",
      "linked db        pathway\n",
      "                 brite\n",
      "                 module\n",
      "                 ko\n",
      "                 genome\n",
      "                 enzyme\n",
      "                 network\n",
      "                 disease\n",
      "                 drug\n",
      "                 ncbi-geneid\n",
      "                 ncbi-proteinid\n",
      "                 uniprot\n",
      "\n"
     ]
    }
   ],
   "source": [
    "kg.keggapi_info(\"hsa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['KEGGapi',\n",
       " 'KEGGerrors',\n",
       " 'KEGGhelpers',\n",
       " 'KEGGpathway',\n",
       " 'KEGGutils',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '__version__',\n",
       " 'connected_components',\n",
       " 'db_categories',\n",
       " 'delete_cached_files',\n",
       " 'download_dir',\n",
       " 'draw',\n",
       " 'get_infos',\n",
       " 'get_nodes_by_nodetype',\n",
       " 'get_organism_codes',\n",
       " 'get_unique_nodetypes',\n",
       " 'graph_measures',\n",
       " 'has_nodetypes',\n",
       " 'is_kegg_up',\n",
       " 'kegg_link_graph',\n",
       " 'keggapi_conv',\n",
       " 'keggapi_find',\n",
       " 'keggapi_get',\n",
       " 'keggapi_info',\n",
       " 'keggapi_link',\n",
       " 'keggapi_list',\n",
       " 'linked_nodes',\n",
       " 'neighbor_graph',\n",
       " 'projected_graph']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(kg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interface for the KEGG API LIST command\n",
      "\n",
      "    See https://www.kegg.jp/kegg/rest/keggapi.html for usage\n",
      "\n",
      "    Parameters\n",
      "    ----------\n",
      "    database : str\n",
      "        Database you wish to obtain a list about, one from pathway | brite | module | ko | genome | <org> | vg | ag | compound | glycan | reaction | rclass | enzyme | network | variant | disease | drug | dgroup | environ | organism | <medicus>\n",
      "    option : str, optional\n",
      "        'xl' option is applicable only to the 'brite' database for listing binary relation files (the default is None)\n",
      "    want_descriptions : bool, optional\n",
      "        If True returns descriptions for each item (the default is False)\n",
      "    force_download : bool, optional\n",
      "        If true replaces any pre-existing downloaded file (the default is False)\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    item, descritpions\n",
      "        lists for items and their descriptions\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(kg.keggapi_list.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Downloading hsa-list from KEGG at http://rest.kegg.jp/list/hsa\n",
      "succesfully downloaded hsa-list\n"
     ]
    }
   ],
   "source": [
    "hsa_list, descriptions = kg.keggapi_list(database = \"hsa\", want_descriptions = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('hsa:90011',\n",
       "  'KIR3DX1, KIR3DL0, LENG12; killer cell immunoglobulin like receptor, three Ig domains X1'),\n",
       " ('hsa:4549', 'RNR1, MTRNR1, MT-RNR1; s-rRNA'),\n",
       " ('hsa:4550', 'RNR2, MTRNR2, MT-RNR2; l-rRNA'),\n",
       " ('hsa:105369154', 'PKD1P6-NPIPP1; PKD1P6-NPIPP1 readthrough'),\n",
       " ('hsa:401983', 'ZNF847P, gm127; zinc finger protein 847, pseudogene'),\n",
       " ('hsa:101926971', 'LINC01284; long intergenic non-protein coding RNA 1284'),\n",
       " ('hsa:574', 'BAGE, BAGE1, CT2.1; B melanoma antigen'),\n",
       " ('hsa:392133',\n",
       "  'OR10AC1, OR10AC1P; olfactory receptor family 10 subfamily AC member 1 (gene/pseudogene)'),\n",
       " ('hsa:2576', 'GAGE4, CT4.4; G antigen 4'),\n",
       " ('hsa:26581', 'DUX5, DUX1; double homeobox 5')]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(zip(hsa_list, descriptions))[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BAGE, BAGE1, CT2.1; B melanoma antigen'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_dictionary = dict(zip(hsa_list, descriptions))\n",
    "\n",
    "genes_dictionary[\"hsa:574\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interface for the KEGG API FIND command\n",
      "    \n",
      "    See https://www.kegg.jp/kegg/rest/keggapi.html for further info\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    database : str\n",
      "        KEGG database you wish to query on, one from pathway | brite | module | ko | genome | genes | <org> | vg | ag | ligand | compound | glycan | reaction | rclass | enzyme | network | variant | disease | drug | dgroup | environ | <medicus>  \n",
      "    query : str\n",
      "        Desired query\n",
      "    option : str, optional\n",
      "        if database is \"compound\" or \"drug\", possible options are \"formula\", \"exact mass\" and \"mol_weight\" (the default is None)\n",
      "    want_descriptions : bool, optional\n",
      "        if set to True returns a list of descriptions for the found items (the default is False,)\n",
      "    verbose : bool, optional\n",
      "        if set to True displays additional messages (the default is False)\n",
      "    force_download : bool, optional\n",
      "        if set to True replaces any previously downloaded file under the same name (the default is False)\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    list, list\n",
      "        list of items and descriptions\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(kg.keggapi_find.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ece:Z1464',\n",
       "  'stx2A; shiga-like toxin II A subunit encoded by bacteriophage BP-933W'),\n",
       " ('ece:Z1465',\n",
       "  'stx2B; shiga-like toxin II B subunit encoded by bacteriophage BP-933W'),\n",
       " ('ece:Z3343',\n",
       "  'stx1B; shiga-like toxin 1 subunit B encoded within prophage CP-933V'),\n",
       " ('ece:Z3344',\n",
       "  'stx1A; shiga-like toxin 1 subunit A encoded within prophage CP-933V'),\n",
       " ('ecs:ECs1205', 'Shiga toxin 2 subunit A'),\n",
       " ('ecs:ECs1206', 'Shiga toxin 2 subunit B'),\n",
       " ('ecs:ECs2973', 'Shiga toxin I subunit B'),\n",
       " ('ecs:ECs2974', 'Shiga toxin I subunit A'),\n",
       " ('ecf:ECH74115_2905', 'shigatoxin 2, subunit B'),\n",
       " ('ecf:ECH74115_2906', 'shiga toxin subunit A')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list, descriptions = kg.keggapi_find(database = \"genes\",\n",
    "                                          query = \"shiga toxin\",\n",
    "                                          force_download = False,\n",
    "                                          want_descriptions = True)\n",
    "\n",
    "list(zip(gene_list, descriptions))[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('cpd:C00493', 'C7H10O5'),\n",
       " ('cpd:C04236', 'C7H10O5'),\n",
       " ('cpd:C16588', 'C7H10O5'),\n",
       " ('cpd:C17696', 'C7H10O5'),\n",
       " ('cpd:C18307', 'C7H10O5'),\n",
       " ('cpd:C18312', 'C7H10O5'),\n",
       " ('cpd:C20961', 'C7H10N2O5'),\n",
       " ('cpd:C21281', 'C7H10O5')]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compounds, formulas = kg.keggapi_find(database = \"compound\",\n",
    "                                          query = \"C7H10O5\",\n",
    "                                          option = \"formula\",\n",
    "                                          force_download = False,\n",
    "                                          want_descriptions = True)\n",
    "\n",
    "list(zip(compounds, formulas))[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interface for the KEGG API GET command\n",
      "\n",
      "    for further info read https://www.kegg.jp/kegg/rest/keggapi.html\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    dbentry : str\n",
      "        KEGG database entry you wish to GET, one from pathway | brite | module | ko | genome | <org> | vg | ag | compound | glycan | reaction | rclass | enzyme | network | variant | disease | drug | dgroup | environ | disease_ja | drug_ja | dgroup_ja | environ_ja | compound_ja\n",
      "    option : str, optional\n",
      "        one from aaseq | ntseq | mol | kcf | image | conf | kgml | json (the default is None])\n",
      "    want_descriptions : bool, optional\n",
      "        if True returns a list of descriptions for the requested items (the default is False)\n",
      "    verbose : bool, optional\n",
      "        is True displays additional messages  (the default is True)\n",
      "    force_download : bool, optional\n",
      "        if set to True replaces any file under the same filename (the default is False)\n",
      "    show_result_image: boo, optional\n",
      "        if set to True shows the downloaded image (the default is True)\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(kg.keggapi_get.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infos on ecs:ECs1206 from KEGG:\n",
      "\n",
      "DEFINITION  (RefSeq) Shiga toxin 2 subunit B\n",
      "ORTHOLOGY   K11007  shiga toxin subunit B\n",
      "ORGANISM    ecs  Escherichia coli O157:H7 Sakai (EHEC)\n"
     ]
    }
   ],
   "source": [
    "kg.keggapi_get(dbentry = \"ecs:ECs1206\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infos on ecs:ECs1206 from KEGG:\n",
      "\n",
      "DEFINITION  (RefSeq) Shiga toxin 2 subunit B\n",
      "ORTHOLOGY   K11007  shiga toxin subunit B\n",
      "ORGANISM    ecs  Escherichia coli O157:H7 Sakai (EHEC)\n"
     ]
    }
   ],
   "source": [
    "kg.keggapi_get(dbentry = \"ecs:ECs1206\", verbose = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Downloading ecs-ECs1206-aaseq from KEGG at http://rest.kegg.jp/get/ecs:ECs1206/aaseq\n",
      "succesfully downloaded ecs-ECs1206-aaseq\n"
     ]
    }
   ],
   "source": [
    "description, sequence = kg.keggapi_get(dbentry = \"ecs:ECs1206\",\n",
    "                                       option = \"aaseq\",\n",
    "                                       want_descriptions = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MKKMFMAVLFALASVNAMAADCAKGKIEFSKYNEDDTFTVKVDGKEYWTSRWNLQPLLQSAQLTGMTVTIKSSTCESGSGFAEVQFNND\n"
     ]
    }
   ],
   "source": [
    "print(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAACVCAYAAABWxTFzAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3X1YVGX+P/D3RxDzEUx5yiTJzHR7MEWjrw+tmoVomK2QhsLid8PL3Ta139a6eRXqr71azaw1u9LyIRUwn0rZkjIN11xLRH9qCrmgq4FP4PqAIeAMfH5/cIaGYQZmmHPmnBk/r+s618zc58y5P8eDn7nnnvucm5gZQgghfFcrvQMQQgihLUn0Qgjh4yTRCyGEj5NEL4QQPk4SvRBC+DhJ9EII4eM0SfREFENEJ4ioiIhma1GHEEII55Da4+iJyA/AvwGMAlAC4ACAScycr2pFQgghnKJFi34QgCJmPsXMNwF8AmCcBvUIIYRwgr8G++wGoNjqdQmAR5p6Q9euXblHjx4ahCKEEL7r4MGDl5g5uLnttEj0ZKesUf8QEaUCSAWAiIgI5OXlaRCKEEL4LiI648x2WnTdlADobvX6TgDnbDdi5g+ZOYqZo4KDm/1AEkII0UJaJPoDAHoRUSQRBQCYCCBLg3qEEEI4QfWuG2Y2E9ELAL4C4AdgFTMfV7seIYQQztGijx7MvB3Adi32LYQQwjVyZawQQvg4SfRCCOHjJNELIYSPk0QvhBA+ThK9EMIQTCYTamtr7a5jZphMJg9H5Dsk0QshDCE0NBTDhg2zm+yzsrIQGBioQ1S+QRK9EMIwcnNzsXjxYr3D8DmS6IUQhjFr1izMnTsXBQUFeofiUyTRCyEMY8CAAZg+fTqSk5NhNpthNpv1DsknSKIXQhjKvHnzcOnSJSxYsAALFizQOxyfoMktEIQQt47q6moEBgaiurq6Re+3neWuXbt2WL58OcaOHQsAeOqpp9yO8VYnLXohhFvWr1+P0aNHg5lbtNgzatQoTJw4ERMnTkRycrIMrXSTtOiFEG7JzMzE888/r/p+LaNv+vTpgzfffFP1/d9KpEUvhBA+ThK9EKLFSktLsX//fk360bt06YIuXbrgnXfewaFDh1Tf/61EEr0QosU2bNiAcePG4bbbbtOsjsTERMTExGi2/1uBJHohRItlZmbiueeea1Q+b948EJFTi8Xly5eRkJBgt57s7GzcuHFDs+PwdS1O9ETUnYhyiKiAiI4T0QylfC4RnSWiw8oSq164xlFVVYWioiKHowaMQGIUWvrPf/6DkydP4vHHH2+0Li0tza1RN9ZycnLwyCOPaHEItwx3WvRmAP+HmfsAiAbwByLqq6x7h5n7KYtPTSn4448/YtiwYejQoQPuu+8+hISE4NlnnzXUJdsSo28w+ofg+vXrER8fD39/bQfvDR06FCdOnEBpaamm9TTH0d01be+saTabHd6Fs7a2Vpehoi1O9Mx8npkPKc+vAygA0E2twIzou+++w6OPPorhw4dj7969qKysxM6dOxEWFobo6GhkZ2frHaLE6AO85UPQUbeN2vz9/TFq1Cjd/y4c3V3T9s6aEREReP/99+3uIyMjQ5+7cLb0Igebr149APwEoBOAuQBOAzgKYBWAzs29f8CAAWx0ZrOZ77zzTp43b57d9S+//DL369fPw1E1JDF6v3379nFQUBC//vrr/N133/HNmzf58OHD/OKLL3KnTp14+/bteofIzMxHjx7liIgIrq2t9Uh9H3/8MU+YMMEjdTnSuXNnbt26Nb/11lsNyrdu3cpt27atfx0eHs5Lliyxu4+1a9c22NZdAPLYmRztzEZN7gDoAOAggGeU16EA/FD3beGvAFY5eF8qgDwAeREREaoduFZOnjzJAPjAgQN212/dupX9/f35xo0bHo7sFxKjd/OmD8HZs2fzn//8Z4/VV1payoGBgXzz5k2P1Wmrc+fO/Morr3D79u05Pz+/vtwbEr1bo26IqDWALQAymPlT5RvCRWauYeZaAB8BGOTgm8SHzBzFzFHBwcHuhOERJ0+eBFD3tcyePn36wGw24/jx454MqwGJ0budOXMGJSUliI21P35h8ODBOHbsGCorKz0cWWOe6raxCA4ORu/evbF3716P1WmP7d01HTl06BA2btzYaNm/f78Ho/1Fi39FobpxUSsBFDDzYqvycGY+r7wcD+CYeyEaQ3h4OADg+vXrCAkJabT+ypUrABwnME+QGL2bKx+CUVFRngytgX379qFDhw548MEHPVrvmDFj8MUXX2D48OEerdfWvHnzcP/992PBggWYM2eO3W0yMzOxadOmRuVmsxmtWnl+VLs7NQ4GMAXACJuhlAuJ6AciOgpgOIBZagSqt759+6JDhw744Ycf7K4/duwYIiIi7CYvT5EYvZv1h6A9RvkQzMzMxOTJkz1e79ixY/H55597vF5blrtrzp8/H0ePHrW7zaJFi/Dzzz83Wj766KNG27799tuIjo7GoEGD8Nprr2kTtDP9O1ov3vBjLDPzM888w4888kijH6BMJhP37t2bk5OT9QnMisTovWpqarhDhw782Wef2V2/YsUK9uTvWTdv3uSamppG5bW1tbr1lYeFhXFRUZEudXfu3Jk3bNhQ/zopKYn79evHmzZtanEffX5+Pj/00ENcWVnJJpOJR4wYwbt27XI6Jniij/5Ws3LlSly7dg1xcXHIz8+HyWTCkSNH8MQTT6Bdu3ZYsmSJ3iFKjF6sVatWeOKJJ/C3v/2t0dh5s9mMt956y6PdFs4OJ/QkS/eNESxevBhnz551686agYGByMzMxG233QZ/f39ERETgyJEjKkZZRxK9C4KCgrB9+3YEBQVh5MiRaNOmDWJiYhAWFobs7Gx06tRJ7xAlRhXs27cPV69e1aVuo30IGm2y7jFjxmDz5s16hwEAqtxw7Y477kDfvnXXmRYUFOCrr75CfHy8WiH+wplmv9aLt3Td2Dp79qzeITRLYnRd3759OTAwkNPS0vjKlSt85coVTeurqKho8PrUqVM8efJkDgsLYyLisLAwnjRpEl+4cEHTOGw5O5zQkz744ANu27Ytp6SkNIhJK9bnxrbrxiImJsbt4ZVHjx7lnj17cnZ2tkvxwVPj6NVYvC3R130+On5tBBJjy+Tk5DCA+iUwMLBB0lfb8ePHuXfv3g7XN/UhuHDhQtXjsWZJbH/605944MCBbDKZmFmfRH/lyhWeMGECP/TQQ3zmzBmeP38+h4aGckxMDO/YsUOTOps7N2rJzc3liIgI/uabb1x+ryR6DRkxQdmSGF2XlZXFjz/+eINEb5v01Uz4q1ev5i5duvCaNWtcfm9xcTEPGDCAExMTubKyUpV4bFkSfUVFBUdGRvIbb7zBzJ5P9N9++y13796dZ8yYwVVVVfXlVVVVvGrVKr7//vv5V7/6Fa9cubLBene4c25cUVJSwt26dXN4AWFzJNFryGgJyh6J0TVLlizhsLAw7tWrl8NEr1bCr6io4KSkJO7Tpw8fP368xTFXVlbyc889x/379+fi4uIW78fali1b6p9bd1Xs2LGDAwIC+MiRIx5J9GazmdPS0jgtLY1DQ0ObvfXD119/zaNHj+aQkBCeO3cul5WVtahetc6Ns2bOnMmtW7fm22+/vX6ZO3eu0++XRK8hIyUoRyRG59TU1PDMmTO5T58+/PXXXzeb5N1N+JbugN/+9reN+uZb6q233uLw8HDeu3dvi/exc+dOHjhwYINz4OxwQrWdPn2a/+d//odHjRrFo0aNcum3iYKCAk5NTeXAwED+3e9+51Ky1uLcaE0SvYaMkKCaIzE2r6Kigp9++mkePnw4X716lf/1r3/xpEmTuH///tyxY0fVE76W3QFffvklBwcH80cffeTS+3Jzc3nkyJENjsfCNtFfunSJg4ODuX///pol+g0bNnBwcDAvWrSIa2tr7d40rbS0tNn9lJWV1ffjP/nkk/zVV181ub2numrUJoleQ3onKGdIjE27ePEiDxw4kJOSkhxe/HP27FnOycnhnJwcXr58Ob/00ks8duxYvvfee9nf39/phO+p7oDCwkLu3bs3//73v6//4dSRgoICfuaZZ+x+cFnYG2WSnp7OABok+sLCQrdj//nnn3nq1Kncq1cvPnjwoMPtVq9ezQsWLHB6v9XV1bx69Wp+4IEHuG/fvrxixYoG/fie7qpRmyR6DUkSVYdeMebn5/Ndd93l8C6RzjCZTHzixAnOysriRYsWcWpqKj/22GMcHh7OQUFB9Qm/tLTUo90B5eXlPHbsWB42bBiXlZU16qs+c+YMp6SksJ+fX6ME36VLFx4yZEj9ts4OJ+zSpQu/++67Lb5l8cGDB7lXr16ckpLCP//8s91trBOyMy16e3bu3MmxsbEcHBzMaWlpfPr0aa/rqrEliV5DrVu3bvDaiElUYrQvJyeHQ0JCeN26dZrVUV5eznl5ebxt2zZ+/PHHecWKFZrVZU9tbS3PmTOHIyIiOCIigi9cuMBlZWU8c+ZMbtu2LUdGRvLo0aP5pZde4g8//JD37NnT4h8vmZmLiop4yJAhPGTIEJda97W1tfz2229zcHCw3Q8UC7X7zn/88UdOTU1lAPzxxx+7vT89+WSiN5lMdu+9UVNT0+y9NyorK7mwsNDtiRJu3rzZYFGLWvExS4xNxRgSEsI5OTmq1decmJgYXrt2rcfqs5aens7p6ekcGBjIW7du5cOHD2t2n//a2lp+9913nW7dX7hwgZ944gl+9NFH+fTp03a3Wb16taZ9561ateKSkhLV9+tJPpnoHV1x5uhqs4KCAh46dCgPHTqU/fz82M/Pj7t27coJCQktvqqua9euDb7utm7dmu+5555Gs844yxKjWvFJjE3H6ImrKa2lp6fzU0895dE6LVatWsWrVq3iMWPGeKxOS+t+8ODBDlv327dv59DQUH799dfZbDY3Wm/dTaNl3/nDDz/MeXl5muzbU5xN9D57rxvreUnVnps0LS0N+fn5yM/Px/fff485c+Zg/vz5WL58eYtjVHvuVInRfox9+vRpcbwtERcXh927d6O8vNyj9QJ185NmZGQgJSXFY3X27NkTe/bsQUJCAqKjo/H3v/+9rkUJoLq6GrNmzcK0adOwefNmzJs3D35+fg3en5+fj/79+6NVq1bIy8tDXl5e/b1g1BYWFoYLFy5osm/DcebTQOtF7RZ9c1OyMbd8WrauXbvysmXLGpWPHDmSn3nmGaf3o+W0cRKjOjGqZdy4cR7vvjl//nz97RvUulrUVdat+3//+9/cr18//s1vfuNwGKqW3TS1tbU8f/78Bl2/U6dOdXk4qtHAV1v09qbosp2eq7kp2QB1p2UrKSnBt99+61LLw5Vp47whRjV4Q4wtER8fb3e2IS198sknGD9+PMaPH482bdp4tG4LS+s+Pj4e999/P6ZPn47NmzcjKCiowXY3btxAcnIyFi5ciD179iApKUn1WIgIK1aswE8//VRfdiu16Fs8lSAAENFpANcB1AAwM3MUEd0OYAOAHgBOA0hg5ivuhfkLe1N02U7P1dyUbIB707J9/PHH2LdvHwCgqKgI+/fvR2hoKF588UWn9+Hq3KlGj7ElU9t5Q4xqiIuLw/Tp01FeXu6xWzBnZGS4dZ90tRARZsyYgfnz52PMmDGN1peVlWHo0KF49NFHkZeXh3bt2mkWS0REBH766Sf06NEDQF2iP3HihGb1GYkaLfrhzNyPmS3/i2YD2MXMvQDsUl6rxt4UXbbTczU3JRvQeFo2ImpysWYymepb2g888AAWLlyI/Px8WCY5d2Zfrkwb5w0xOrNPI8Soh44dO2LEiBHYtm2bR+orLCxEcXExRowYgREjRnikzuZERkbi3LlzjcpXr16NV199FatXr9Y0yQO/JHoLadG7ZxyAXyvP1wDYDeDPGtTjkPW8pD179rS7je3cpKz8YOSM559/HtOmTXO43pl9NRejvblTjR6jK/HpFaNe4uPjsWHDBkyZMkXzujIyMjBp0iRdJqF2JDw8HOfPn29UnpKSUv/BrjXbRO8oJl/k7l8CA9hBRAeJKFUpC2Xm8wCgPHr8f5jtlGy2CUOPadlsGW3aOHskRvV4cvRNZmYmnnvuOc3rccUdd9xht0XvqSQP3NotencT/WBm7g9gNIA/ENEwZ99IRKlElEdEeWVlZW6G0Zj1lGxGmZatqRiNGJ/EqJ6OHTti+PDhyMrK0rSevLw8MDMGDhyoaT2uMkLrWVr0LcTM55THUgCfARgE4CIRhQOA8ljq4L0fMnMUM0dp8aluPS+pUecmNfrcqRKjuhISErBx40ZN68jIyEBiYqKmdbSEoxa9J9km+vbt28PPz0+Xaxw8zpkxmPYWAO0BdLR6vg9ADIC3AMxWymcDWNjcvjx1rxujzU1qy+jxMUuM7igvL+eOHTvytWvXNNl/TU0Nh4WF8YkTJzTZvzuysrI8eoWuPVevXuWOHTs2KLvnnnsM+e/lLHhgHH0ogL1EdARALoAvmPlLAH8DMIqICgGMUl4bwh133KHq/mxHkbhL7fgAidFItO6+ycnJQbdu3XDvvfdqsn93GKFFHxgYCCLC1atX68tulX76Fo+6YeZTAB6yU/5fACPdCUoIX2W5eGry5Mmq79uo3TaAMRI98Ev3jeWirVuln57YxSFxWoiKiuK8vDy9w3AZEbk8pNDTJEZjKS8vR/fu3VFcXKzqbwfV1dUICwtDfn5+/bUFRlJbW4uAgABUV1c3ur+Np5lMJvj5+TUafsrMMJvNaN26tU6RuY6IDvIv1zA5ZJyBtkIIITQhiV4ID+rUqRN+/etfq95P//nnn2PAgAGGbM0Dddc7hISEGKI/PDQ0FMOGDUNtbW2D8qysLAQGBuoUlbYk0QvhYVrc5MyIF0nZMlJ/eG5uLhYvXqx3GB4jiV4ID4uPj0dFRYVT9wZydvniiy+QmJho6N86jPKDLADMmjULc+fORUFBgd6heIQkeiE86ObNm6iursbOnTtVndOhqqoKxcXFePLJJ3Hq1Cm9D7OeyWSq7yL5xz/+gbi4OAB1P3yaTCbd4howYACmT5+O5ORkmM1m3eLwFEn0brjnnnv0DqFZEqOxbNmyBd27d8e4cePw3//+V5V9MjOWLVuG6OhojBkzBpGRkYZp2Ru5P3zevHm4dOkSFixYoGscnnDLJ3qz2dzojxCoGw7WVIujqqoK2dnZmv+HMpvNXhFjU/EZIUajmDRpEoqLixEfH4+2bduqsk8iQkBAAPbv348ZM2bg5MmTmDVrFmpqalBTU6NKHe4wan94u3btsHz5csyfPx9Hjx7VOxxN3fKJPiIiAu+//36j8oyMDLstjh9//BHDhg1Dhw4dcN999yEkJATPPvusZn19lvvRGz3GpuIzQoxGsH79eqSnpwMAJk+erOr911NSUnD33XdjyZIliI6ORmRkJFq1amWIWxU70x9+5MgRD0b0i1GjRmHixIlITk7WtStJa/r/FXgRLSfKlhiNFaPamBmtWrXCpk2b0L17d8yZM0fV/RMRVq9eXT+15owZM+rX5efnq1qXq5zpD4+NjUVMTAx2797t2eAALF68GGfPnjXEjFyaUfMHoZYunrqpmT1qTTiu1STU4eHhXhFjU/EZIUYjuXbtGn/33Xeq77eqqopra2uZmbmwsJCHDBnCQ4YM4eHDh7PJZFK9Pmd07tyZN2zYwBUVFRwZGclvvPEGMzNv3bq1wd9FdXU1r1y5ku+9914eNGgQb9myhWtqahpM5q12TNbS09MZQIOYvAGcvKmZFjNMeR3LhOPWXJ1wfPDgwXjnnXdQWVmpWt+rN8XoTHx6x6iXqqoqTJ06FbGxsYiLi0OnTp0QHR2tej1t2rQBM2PJkiWYP38+XnvtNQCon4O3oqIC7du3V71eZ1j6w8eOHYunnnqq0fqAgABMnToVKSkp2Lp1KxYsWIC//OUvAIBXXnkFU6ZMQUBAQIvrN5vN8Pd3nO4SExORnp6Of/7zny2uw9Cc+TTQetG7RR8QEMDt27dvsLRp06bBp/uOHTsYAF+8eNHufk6cOMEA+MCBA6rH5w0xNhWfEWLUk8lk4nXr1nFcXBx36tSJ4+LiuLi4WJO6rl+/zk8++SQXFRXVl1la96mpqfUtfk+xbT0nJSVxv379eNOmTc22nnfv3s27d+/mmJgYDg8P54ULF3J5ebnLMZw7d46HDBni8vu8ATxwm2KfocaE41pPQm30GJ2JT+8Y9VBVVYWkpCQAwLp16+pH3Kh9a2aLDh06IDs7u/6HWcuPsxMmTMCyZcs0q9dZjvrDc3NzG2372GOP4bHHHkN2dja+/PJLHD58GD169MCrr76KiooKp+r75ptv0L9/f8TExNwyI7vskUTvJOtJqO0xwiTUEqPx+Pv7IzY2tv5H2ClTpoCINL1nPjMjNjYWGzdubPDjrOUuoYcPH9as7uZ06dIF77zzDg4dOtSgPD4+HiNHjsSuXbvsvu/BBx9ERkYGDh48iPLycixdurTJepgZf/3rX5GYmIiMjAzMmTNH9w85PUmid5I3TEItMRoLM8Pf3x+TJ0/Gtm3b6lvze/fu1TTptGrVCmlpadi7dy/27t2Lnj171q8jItx3330oKipy2MJt6XUbzkpMTERMTEyDspMnTyI5ORkvvPACBg4ciM8++6y+28Fajx49sHTpUvzxj390uP/Lly8jNjYWX375JQ4dOoQRI0a4HbO3k0TvAm+YhFpiNI59+/Zh2rRpSE9PR3l5OTp16oTJkyfjgw8+0Lxuez/2OnvtgqvXbTTl8uXLSEhIaFSenZ2NGzdu1L/29/dHUlIS8vPzMWfOHLz55pvo27cv+vbtizVr1jQalunoGoTc3Fw8/PDDeOCBB5CTk2PYu3l6WosTPRH1JqLDVks5Ec0korlEdNaq3P7wCi/kDZNQS4zqcufK5AEDBuDll19GYmKi7sfkLdcuEBGefvpp5ObmYunSpVi6dCnWrVuHyMhIvPfee6isrHT43qVLl2LMmDFYsmQJFi5c2OQom1uOM7/YNrcA8ANwAcBdAOYC+JMr79dz1I07jDoJtTWJ0T2uXMdgUVBQwEOHDmU/Pz/28/Pjrl27ckJCAufn53si5EZcvXbB1eP1hAMHDvD48eM5ODiY33jjDb569Wr9uuvXr3NCQgL369ePT506pUt8eoGHR92MBHCSmc+otD+vYNRJqK1JjJ5lxJazM9cuHDt2rEFr2XJdhPVi77oIT4mKisKnn36KPXv2oLCwEJGRkZg9ezZOnz6NAQMGICgoCN9//z0iIyN1i9HQnPk0aG4BsArAC8rzuQBOAziqlHdu7v3e2qIXvs8brkxujqvXLjh73Yaezpw5wy+88AKPHj2a165dq3c4uoGnrowlogAAcQD+ohR9AOD/AmDl8W0AU+28LxVAKuA7Y6aF7zL6lclNsb52wd6wVXvXLixatKjRyJZ169Zh2rRpGkbqvIiICLz33ns4c+YM7rrrLr3DMTw1fq0YDeAQM18EAMsjABDRRwA+t/cmZv4QwIcAEBUVdeteySC8QmZmZqPp/8xmc4O7Q548eRKA44ZLnz59YDabcfz4cURFRWkXrA3raxesh1paePO1C5LknaNGH/0kAOstL4jIejzTeADHVKhDCF0Z/crkptxK1y4I+9xK9ETUDsAoAJ9aFS8koh+I6CiA4QBmuVOHEN7CyFf93irXLgj73Er0zHyDmbsw8zWrsinM/AAzP8jMccxsjGnfhdCYkVvO3nTtglCfXBkrhIqM3HKOjIzEunXrcP78eZSUlOD8+fPIzMxEaGhog+3OnTtn9xYDU6ZMaXA1q/AekuiFUJG3tJx96doF0Tyy/Yqph6ioKM7Ly9M7DCFUd+7cOUmqQjNEdJCZmx3CJS16ITQkSV4YgSR6IYTwcZLohRDCx0miF0IIHyeJXgghfJwkeiGE8HGS6IUQwsdJohdCCB8niV4IIXycJHohhPBxkuiFEMLHqTHDlMcRUYPX1vfrsV5n7z4+Ta3XYp0QQujN6xI9ETlM7PbWOftai3VCCGEEzXbdENEqIiolomNWZbcT0ddEVKg8dlbKiYiWEFERER0lov5qBmsvieqdVPWuXwghmuNMH/3HAGJsymYD2MXMvQDsUl4DdROF91KWVAAfqBOmd5DWvBDCiJpN9My8B8Blm+JxANYoz9cAeNqqfC3X+R5AkM1k4T5LkrwQwqhaOuom1DIXrPJome24G4Biq+1KlDKfJkleCGFkag+vJDtldjMgEaUSUR4R5ZWVlTm1c2ZuNOLG9rWnNfXjsBBCGEFLE/1FS5eM8liqlJcA6G613Z0AztnbATN/yMxRzBwVHBzsdMWWZG9ZrJNsU+vceW9z+7VeJ4QQRtPSRJ8FIFl5ngxgm1V5kjL6JhrANUsXj5qYuX5xZZ0773W0zrq8qXqFEEIvzY6jJ6L1AH4NoCsRlQBIA/A3ABuJ6H8B/AQgXtl8O4BYAEUAbgBI0SBmIYQQLmg20TPzJAerRtrZlgH8wd2ghBBCqEfudSOEED5OEr0QQvg4SfRCCOHjJNELIYSPk0QvhBA+ThK9EEL4OEn0Qgjh48gIV3ISURmACgCX9I5FJV0hx2JEcizGJMfScncxc7P3kDFEogcAIspj5ii941CDHIsxybEYkxyL9qTrRgghfJwkeiGE8HFGSvQf6h2AiuRYjEmOxZjkWDRmmD56IYQQ2jBSi14IIYQGdE/0RBRDRCeIqIiIZusdj6uI6DQR/UBEh4koTym7nYi+JqJC5bGz3nHaQ0SriKiUiI5ZldmNXZlMZolyno4SUX/9Im/MwbHMJaKzyrk5TESxVuv+ohzLCSJ6Up+o7SOi7kSUQ0QFRHSciGYo5V53bpo4Fq87N0R0GxHlEtER5VjmKeWRRLRfOS8biChAKW+jvC5S1vfQLXh7MyR5agHgB+AkgLsBBAA4AqCvnjG14BhOA+hqU7YQwGzl+WwAC/SO00HswwD0B3CsudhRN6FMNurmBY4GsF/v+J04lrkA/mRn277K31obAJHK36Cf3sdgFV84gP7K844A/q3E7HXnpolj8bpzo/z7dlCetwawX/n33ghgolK+DMB05flOJaYEAAADCUlEQVTvASxTnk8EsEGv2PVu0Q8CUMTMp5j5JoBPAIzTOSY1jAOwRnm+BsDTOsbiEDPvAXDZpthR7OMArOU63wMIsswbbAQOjsWRcQA+YeZqZv4P6mZEG6RZcC5i5vPMfEh5fh1AAYBu8MJz08SxOGLYc6P8+/6svGytLAxgBIDNSrntebGcr80ARpJOE0vrnei7ASi2el2Cpv8IjIgB7CCig0SUqpSFsjJXrvIYolt0rnMUu7eeqxeU7oxVVl1oXnMsytf9h1HXevTqc2NzLIAXnhsi8iOiwwBKAXyNum8cV5nZrGxiHW/9sSjrrwHo4tmI6+id6O19unnbMKDBzNwfwGgAfyCiYXoHpBFvPFcfAOgJoB+A8wDeVsq94liIqAOALQBmMnN5U5vaKTPU8dg5Fq88N8xcw8z9ANyJum8afextpjwa5lj0TvQlALpbvb4TwDmdYmkRZj6nPJYC+Ax1J/+i5auz8liqX4QucxS7150rZr6o/MesBfARfukCMPyxEFFr1CXGDGb+VCn2ynNj71i8+dwAADNfBbAbdX30QURkmX/bOt76Y1HWB8L57kVV6Z3oDwDopfxqHYC6HyyydI7JaUTUnog6Wp4DeALAMdQdQ7KyWTKAbfpE2CKOYs8CkKSM8IgGcM3SjWBUNv3U41F3boC6Y5mojIqIBNALQK6n43NE6cddCaCAmRdbrfK6c+PoWLzx3BBRMBEFKc/bAngcdb855ACYoGxme14s52sCgG9Y+WXW4wzwS3Ys6n6JPwlgjt7xuBj73agbIXAEwHFL/Kjrh9sFoFB5vF3vWB3Evx51X5tNqGt9/K+j2FH3NfR95Tz9ACBK7/idOJZ1SqxHUfefLtxq+znKsZwAMFrv+G2OZQjqvuIfBXBYWWK98dw0cSxed24APAjg/ykxHwPwulJ+N+o+jIoAbALQRim/TXldpKy/W6/Y5cpYIYTwcXp33QghhNCYJHohhPBxkuiFEMLHSaIXQggfJ4leCCF8nCR6IYTwcZLohRDCx0miF0IIH/f/Adl0dKteweKtAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "img = kg.keggapi_get(dbentry = \"C00002\",\n",
    "              option = \"image\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> deleting the following files from ./kegg_downloads/\n",
      "./kegg_downloads\\C00002-image.gif\n",
      "./kegg_downloads\\compound-C7H10O5-formula\n",
      "./kegg_downloads\\eco-ncbi-geneid-19831932-link\n",
      "./kegg_downloads\\ecs-ECs1206-aaseq\n",
      "./kegg_downloads\\ecs-ECs1206-description\n",
      "./kegg_downloads\\enzyme-hsa-link\n",
      "./kegg_downloads\\genes-ncbi-geneid-19831932-link\n",
      "./kegg_downloads\\genes-shiga-toxin\n",
      "./kegg_downloads\\hsa-info\n",
      "./kegg_downloads\\hsa-list\n",
      "./kegg_downloads\\hsa05130-description\n",
      "./kegg_downloads\\hsa05130-image.png\n",
      "./kegg_downloads\\hsa05130-kgml\n",
      "./kegg_downloads\\ncbi-geneid-hsa-link\n",
      "./kegg_downloads\\organism-code-list\n"
     ]
    }
   ],
   "source": [
    "kg.delete_cached_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infos on hsa05130 from KEGG:\n",
      "\n",
      "NAME        Pathogenic Escherichia coli infection - Homo sapiens (human)\n",
      "DESCRIPTION Enteropathogenic E. coli (EPEC) and enterohemorrhagic E. coli (EHEC) are closely related pathogenic strains of Escherichia coli. The hallmark of EPEC/EHEC infections [DS:H00278 H00277] is induction of attaching and effacing (A/E) lesions that damage intestinal epithelial cells. The capacity to form A/E lesions is encoded mainly by the locus of enterocyte effacement (LEE) pathogenicity island. Tir, Map, EspF, EspG are known LEE-encoded effector proteins secreted via the type III secretion system, which is also LEE-encoded, into the host cell. EPEC and EHEC Tir's link the extracellular bacterium to the cell cytoskeleton. Map and EspF are involved in mitochondrion membrane permeabilization. EspG interacts with tubulins and stimulates microtubule destabilization. LEE-encoded adhesin or intimin (Eae) is exported via the general secretory pathway to the periplasm, where it is inserted into the outer membrane. In addition to Tir, two potential host cell-carried intimin receptors, beta1 integrin (ITGB1) and nucleolin (NCL), have so far been identified. The distinguishing feature of EHEC is the elaboration of Shiga-like toxin (Stx). Stx cleaves ribosomal RNA, thereby disrupting protein synthesis and killing the intoxicated epithelial or endothelial cells.\n",
      "CLASS       Human Diseases; Infectious diseases: Bacterial\n"
     ]
    }
   ],
   "source": [
    "kg.keggapi_get(dbentry = \"hsa05130\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Downloading hsa05130-image from KEGG at http://rest.kegg.jp/get/hsa05130/image\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support.' +\n",
       "              'Please try Chrome, Safari or Firefox â‰¥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            if (mpl.ratio != 1) {\n",
       "                fig.send_message(\"set_dpi_ratio\", {'dpi_ratio': mpl.ratio});\n",
       "            }\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        fig.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var backingStore = this.context.backingStorePixelRatio ||\n",
       "\tthis.context.webkitBackingStorePixelRatio ||\n",
       "\tthis.context.mozBackingStorePixelRatio ||\n",
       "\tthis.context.msBackingStorePixelRatio ||\n",
       "\tthis.context.oBackingStorePixelRatio ||\n",
       "\tthis.context.backingStorePixelRatio || 1;\n",
       "\n",
       "    mpl.ratio = (window.devicePixelRatio || 1) / backingStore;\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width * mpl.ratio);\n",
       "        canvas.attr('height', height * mpl.ratio);\n",
       "        canvas.attr('style', 'width: ' + width + 'px; height: ' + height + 'px;');\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option)\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'] / mpl.ratio;\n",
       "    var y0 = (fig.canvas.height - msg['y0']) / mpl.ratio;\n",
       "    var x1 = msg['x1'] / mpl.ratio;\n",
       "    var y1 = (fig.canvas.height - msg['y1']) / mpl.ratio;\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width, fig.canvas.height);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x * mpl.ratio;\n",
       "    var y = canvas_pos.y * mpl.ratio;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"jpeg\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\", \"tif\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overridden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    var width = fig.canvas.width/mpl.ratio\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\" width=\"' + width + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var width = this.canvas.width/mpl.ratio\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\" width=\"' + width + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        event.shiftKey = false;\n",
       "        // Send a \"J\" for go to next cell\n",
       "        event.which = 74;\n",
       "        event.keyCode = 74;\n",
       "        manager.command_mode();\n",
       "        manager.handle_keydown(event);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA2gAAAKOCAYAAADTfDoWAAAIuklEQVR4nO3BMQEAAADCoPVPbQdvoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB4D00wAAW950bYAAAAASUVORK5CYII=\" width=\"639.4666666666667\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib notebook\n",
    "e_coli_infection_pathway_img = kg.keggapi_get(dbentry = \"hsa05130\", option = \"image\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "kgml = kg.keggapi_get(dbentry = \"hsa05130\", option = \"kgml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pathway Title:\n",
      "Pathogenic Escherichia coli infection\n",
      "Pathway link:\n",
      "http://www.kegg.jp/kegg-bin/show_pathway?hsa05130\n",
      "Pathway image link:\n",
      "http://www.kegg.jp/kegg/pathway/hsa/hsa05130.png\n"
     ]
    }
   ],
   "source": [
    "print(\"Pathway Title:\")\n",
    "print(kgml.getroot().get('title'))\n",
    "print(\"Pathway link:\")\n",
    "print(kgml.getroot().get('link'))\n",
    "print(\"Pathway image link:\")\n",
    "print(kgml.getroot().get('image'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interface for the KEGG API LINK command \n",
      "\n",
      "    for further info see \n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    source : [type]\n",
      "        [description]\n",
      "    target : [type]\n",
      "        [description]\n",
      "    verbose : bool, optional\n",
      "        [description] (the default is True, which [default_description])\n",
      "    force_download : bool, optional\n",
      "        [description] (the default is False, which [default_description])\n",
      "    \n",
      "    Raises\n",
      "    ------\n",
      "    KEGGKeyError\n",
      "        [description]\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    [type]\n",
      "        [description]\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(kg.keggapi_link.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Downloading enzyme-hsa-link from KEGG at http://rest.kegg.jp/link/enzyme/hsa\n",
      "succesfully downloaded enzyme-hsa-link\n"
     ]
    }
   ],
   "source": [
    "genes, enzymes = kg.keggapi_link(\"hsa\", \"enzyme\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('hsa:9344', 'ec:2.7.11.1'),\n",
       " ('hsa:5894', 'ec:2.7.11.1'),\n",
       " ('hsa:673', 'ec:2.7.11.1'),\n",
       " ('hsa:5607', 'ec:2.7.12.2'),\n",
       " ('hsa:5598', 'ec:2.7.11.24'),\n",
       " ('hsa:9020', 'ec:2.7.11.25'),\n",
       " ('hsa:51701', 'ec:2.7.11.24'),\n",
       " ('hsa:64840', 'ec:2.3.1.250'),\n",
       " ('hsa:495', 'ec:7.2.2.19'),\n",
       " ('hsa:496', 'ec:7.2.2.19')]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(zip(genes, enzymes))[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:> File eco-ncbi-geneid-link already present in ./kegg_downloads/\n",
      "reading from chached file...\n"
     ]
    }
   ],
   "source": [
    "ncbi_genes, kegg_genes_eco = kg.keggapi_conv(\"ncbi-geneid\", \"eco\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ncbi-geneid:944742', 'eco:b0001'),\n",
       " ('ncbi-geneid:945803', 'eco:b0002'),\n",
       " ('ncbi-geneid:947498', 'eco:b0003'),\n",
       " ('ncbi-geneid:945198', 'eco:b0004'),\n",
       " ('ncbi-geneid:944747', 'eco:b0005'),\n",
       " ('ncbi-geneid:944749', 'eco:b0006'),\n",
       " ('ncbi-geneid:944745', 'eco:b0007'),\n",
       " ('ncbi-geneid:944748', 'eco:b0008'),\n",
       " ('ncbi-geneid:944760', 'eco:b0009'),\n",
       " ('ncbi-geneid:944792', 'eco:b0010')]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(zip(ncbi_genes, kegg_genes_eco))[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "http://rest.kegg.jp/conv/genes/ncbi-geneid:19831932"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:> File genes-ncbi-geneid-19831932-link already present in ./kegg_downloads/\n",
      "reading from chached file...\n"
     ]
    }
   ],
   "source": [
    "ncbi, kegg_eco = kg.keggapi_conv(\"ncbi-geneid:19831932\", \"genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ncbi-geneid:19831932 NCBI gene is called lpn:lpg0366 in KEGG \n"
     ]
    }
   ],
   "source": [
    "print(\"{} NCBI gene is called {} in KEGG \".format(ncbi[0], kegg_eco[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'KEGGutils' has no attribute 'keggapi_DDI'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-44-a54a8a7754be>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mddi_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeggapi_DDI\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"D00564\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m: module 'KEGGutils' has no attribute 'keggapi_DDI'"
     ]
    }
   ],
   "source": [
    "ddi_list = kg.keggapi_DDI(\"D00564\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
